Assertion hit at mediator state machine during mediation

## Problem Definition

Mediating 5 "parallel" transfers (new mediated transfer requests arriving before the previous one for that pair were finished), **at least one** of them triggered assertions in the state machine, which repeated on every new mined block, and changed but kept occurring after expiration block was hit.

The mediated transfer logs I could find before the error started, is as follow:
- My address: `0xb0d5eeaeb7b084a2ba92c24b79f4f2d079a3062b`
- Token: `RTT 0x0f114a1e9db192502e7856309cc899952b3db1ed`
- Originator channel: `0x0bd6ca510d42773f4943f8c00da94b7acc798e4c`
- Originator partner: `0x008311788dff04e3c0426a3eb19622b2035ba49d`
- Target channel: `0xc5ba20b92c3066421868cf40e919e577836ca016`
- Target partner: `0x02f4b6bc65561a792836212ebc54434db0ab759a` (echo node)
- Amount: 7


### Mediation logs:
```
DEBUG:raiden.tasks	new block timestamp=1505316501.39 number=1666736
DEBUG:raiden.network.protocol.ping	PING RECEIVED node=b0d5eeae message=<Ping [msghash=f569c689]> sender=02f4b6bc echohash=664931e8
DEBUG:raiden.network.protocol	ACK RECEIVED node=b0d5eeae receiver=00831178 echohash=df4c1d2c
DEBUG:raiden.network.protocol	ACK RECEIVED node=b0d5eeae receiver=02f4b6bc echohash=a7f97bf8
INFO:raiden.network.protocol	MESSAGE RECEIVED node=b0d5eeae message=<MediatedTransfer [channel:0bd6ca51 nonce:2 transferred_amount:0 locksroot:a7e100c9 hash:1e23b31a id:13328499716505091091 hashlock:b5f8922b expiration:1666826 amount:7]> message_sender=00831178 echohash=a0754b33
DEBUG:raiden.channel.netting_channel	REGISTERED LOCK node=b0d5eeae currentlocksroot=48d6154f lock_expiration=1666826 to=b0d5eeae lock_amount=7 lock_hashlock=b5f8922b from_=00831178 lockhashes=['48d6154f'] lockhash=c81f970e
DEBUG:raiden.channel.netting_channel	REGISTERED TRANSFER node=b0d5eeae nonce=None transfer=<MediatedTransfer [channel:0bd6ca51 nonce:2 transferred_amount:0 locksroot:a7e100c9 hash:1e23b31a id:13328499716505091091 hashlock:b5f8922b expiration:1666826 amount:7]> to=b0d5eeae transferred_amount=0 from_=00831178 current_locksroot=a7e100c9
DEBUG:raiden.channel.netting_channel	REGISTERED LOCK node=b0d5eeae currentlocksroot=cbae5128 lock_expiration=1666814 to=02f4b6bc lock_amount=7 lock_hashlock=b5f8922b from_=b0d5eeae lockhashes=['cbae5128'] lockhash=63e551fa
DEBUG:raiden.channel.netting_channel	REGISTERED TRANSFER node=b0d5eeae nonce=32 transfer=<MediatedTransfer [channel:c5ba20b9 nonce:36 transferred_amount:94 locksroot:488f7684 hash:5a1695de id:13328499716505091091 hashlock:b5f8922b expiration:1666814 amount:7]> to=02f4b6bc transferred_amount=94 from_=b0d5eeae current_locksroot=488f7684
DEBUG:raiden.network.protocol	SENDING MESSAGE node=b0d5eeae to=02f4b6bc message=<MediatedTransfer [channel:c5ba20b9 nonce:36 transferred_amount:94 locksroot:488f7684 hash:5a1695de id:13328499716505091091 hashlock:b5f8922b expiration:1666814 amount:7]> echohash=d7d9b817
DEBUG:raiden.network.protocol	SENDING ACK node=b0d5eeae to=00831178 echohash=a0754b33
DEBUG:raiden.network.protocol	ACK RECEIVED node=b0d5eeae receiver=02f4b6bc echohash=d7d9b817
...
DEBUG:raiden.tasks	new block timestamp=1505317331.21 number=1666806
INFO:raiden.network.protocol	MESSAGE RECEIVED node=b0d5eeae message=<RevealSecret [hashlock:b5f8922b hash:fdb99a97]> message_sender=02f4b6bc echohash=052b62a8
DEBUG:raiden.channel.netting_channel	SECRET REGISTERED node=b0d5eeae hashlock=b5f8922b token=0f114a1e to=00831178 amount=7 from_=b0d5eeae
DEBUG:raiden.network.protocol	SENDING MESSAGE node=b0d5eeae to=00831178 message=<RevealSecret [hashlock:b5f8922b hash:8eb71b0d]> echohash=2dc5c410
DEBUG:raiden.channel.netting_channel	SECRET REGISTERED node=b0d5eeae hashlock=b5f8922b token=0f114a1e to=b0d5eeae amount=7 from_=02f4b6bc
DEBUG:raiden.network.protocol	SENDING MESSAGE node=b0d5eeae to=02f4b6bc message=<RevealSecret [hashlock:b5f8922b hash:8eb71b0d]> echohash=01bc880d
DEBUG:raiden.channel.netting_channel	SECRET REGISTERED node=b0d5eeae hashlock=b5f8922b token=0f114a1e to=00831178 amount=7 from_=b0d5eeae
DEBUG:raiden.channel.netting_channel	REGISTERED TRANSFER node=b0d5eeae nonce=32 transfer=<Secret [channel:c5ba20b9 nonce:41 transferred_amount:108 locksroot:9573d5e8 hash:8c8a5f6c hashlock:b5f8922b]> to=02f4b6bc transferred_amount=108 from_=b0d5eeae current_locksroot=9573d5e8
DEBUG:raiden.network.protocol	SENDING MESSAGE node=b0d5eeae to=02f4b6bc message=<Secret [channel:c5ba20b9 nonce:41 transferred_amount:108 locksroot:9573d5e8 hash:8c8a5f6c hashlock:b5f8922b]> echohash=746e2093
DEBUG:raiden.network.protocol	SENDING ACK node=b0d5eeae to=02f4b6bc echohash=052b62a8
DEBUG:raiden.network.protocol	ACK RECEIVED node=b0d5eeae receiver=02f4b6bc echohash=0b6c8ed9
```

### Initial assertion, repeated on every new block
```
DEBUG:raiden.tasks	new block timestamp=1505317390.52 number=1666810
ERROR:raiden.tasks	unexpected exception on alarm 
Traceback (most recent call last):
  File "/apps/raiden/raiden/tasks.py", line 107, in poll_for_new_block
    result = callback(current_block)
  File "/apps/raiden/raiden/raiden_service.py", line 369, in set_block_number
    self.state_machine_event_handler.log_and_dispatch_to_all_tasks(state_change)
  File "/apps/raiden/raiden/event_handler.py", line 63, in log_and_dispatch_to_all_tasks
    events = self.dispatch(manager, state_change)
  File "/apps/raiden/raiden/event_handler.py", line 95, in dispatch
    all_events = state_manager.dispatch(state_change)
  File "/apps/raiden/raiden/transfer/architecture.py", line 130, in dispatch
    state_change,
  File "/apps/raiden/raiden/transfer/mediated_transfer/mediator.py", line 941, in state_transition
    iteration = handle_block(state, state_change)
  File "/apps/raiden/raiden/transfer/mediated_transfer/mediator.py", line 708, in handle_block
    block_number,
  File "/apps/raiden/raiden/transfer/mediated_transfer/mediator.py", line 394, in set_expired_pairs
    assert pair.payee_state not in STATE_TRANSFER_PAID
AssertionError
```

### Assertion after expiration, repeated every new block
```
DEBUG:raiden.tasks	new block timestamp=1505317776.99 number=1666829
ERROR:raiden.tasks	unexpected exception on alarm 
Traceback (most recent call last):
  File "/apps/raiden/raiden/tasks.py", line 107, in poll_for_new_block
    result = callback(current_block)
  File "/apps/raiden/raiden/raiden_service.py", line 369, in set_block_number
    self.state_machine_event_handler.log_and_dispatch_to_all_tasks(state_change)
  File "/apps/raiden/raiden/event_handler.py", line 63, in log_and_dispatch_to_all_tasks
    events = self.dispatch(manager, state_change)
  File "/apps/raiden/raiden/event_handler.py", line 95, in dispatch
    all_events = state_manager.dispatch(state_change)
  File "/apps/raiden/raiden/transfer/architecture.py", line 130, in dispatch
    state_change,
  File "/apps/raiden/raiden/transfer/mediated_transfer/mediator.py", line 941, in state_transition
    iteration = handle_block(state, state_change)
  File "/apps/raiden/raiden/transfer/mediated_transfer/mediator.py", line 708, in handle_block
    block_number,
  File "/apps/raiden/raiden/transfer/mediated_transfer/mediator.py", line 381, in set_expired_pairs
    assert pair.payee_state == 'payee_expired'
AssertionError
```

### Changing assertion above to give more information about the `pair` object (linebreaks added):
```
...
     assert pair.payee_state == 'payee_expired', repr(pair)                         
AssertionError: <MediationPairState payer_route:<RouteState opened hop:00831178 available_balance:31 settle:90 reveal:10> 
    payer_transfer:<HashTimeLocked id=13328499716505091091 amount=7 token=0f114a1e9db192502e7856309cc899952b3db1ed target=02f4b6bc65561a792836212ebc54434db0ab759a expire=1666826 hashlock=b5f8922b3bd3db744de9a1fdb98d7158a23a560c0e215192e4ff594f317aa181>
    payer_state:payer_waiting_close payee_route:<RouteState opened hop:02f4b6bc available_balance:221 settle:90 reveal:10> 
    payee_transfer:<HashTimeLocked id=13328499716505091091 amount=7 token=0f114a1e9db192502e7856309cc899952b3db1ed target=02f4b6bc65561a792836212ebc54434db0ab759a expire=1666814 hashlock=b5f8922b3bd3db744de9a1fdb98d7158a23a560c0e215192e4ff594f317aa181> 
    payee_state:payee_balance_proof>
```

## Solution

Investigate why a `STATE_TRANSFER_PAID` pair ended up in pending pairs, and handle it.

## System Description
Master 7f5cf59887f7c172353ad8f3f7abcdb125346c66 over `geth 1.6.7-stable-ab5646c5`

--- a/raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py
+++ b/raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py
@@ -1,0 +1,110 @@
+# -*- coding: utf-8 -*-
+# pylint: disable=invalid-name,too-many-locals,too-many-arguments,too-many-lines
+from raiden.transfer.mediated_transfer import mediator
+from raiden.transfer.mediated_transfer.state import MediatorTransferState
+from raiden.transfer.mediated_transfer.state_change import ReceiveSecretReveal
+from raiden.transfer.mediated_transfer.events import SendMediatedTransfer
+from raiden.transfer.state_change import Block
+from raiden.tests.utils import factories
+from raiden.tests.utils.events import must_contain_entry
+from raiden.tests.utils.factories import (
+    HOP1,
+    HOP2,
+    UNIT_HASHLOCK,
+    UNIT_SECRET,
+    UNIT_TOKEN_ADDRESS,
+    UNIT_TRANSFER_SENDER,
+)
+
+
+def test_payer_enter_danger_zone_with_transfer_payed():
+    """ A mediator may have paid the next hop (payee), and didn't get payed by
+    the previous hop (payer).
+
+    When this happens, an assertion must not be hit, because it means the
+    transfer must be withdrawn on-chain.
+
+    Issue: https://github.com/raiden-network/raiden/issues/1013
+    """
+    amount = 10
+    block_number = 5
+    target = HOP2
+    expiration = 30
+
+    payer_channel = factories.make_channel(
+        partner_balance=amount,
+        partner_address=UNIT_TRANSFER_SENDER,
+        token_address=UNIT_TOKEN_ADDRESS,
+    )
+
+    payer_transfer = factories.make_signed_transfer_for(
+        payer_channel,
+        amount,
+        HOP1,
+        target,
+        expiration,
+        UNIT_SECRET,
+    )
+
+    channel1 = factories.make_channel(
+        our_balance=amount,
+        token_address=UNIT_TOKEN_ADDRESS,
+    )
+    channelmap = {
+        channel1.identifier: channel1,
+        payer_channel.identifier: payer_channel,
+    }
+    possible_routes = [factories.route_from_channel(channel1)]
+
+    mediator_state = MediatorTransferState(UNIT_HASHLOCK)
+    initial_iteration = mediator.mediate_transfer(
+        mediator_state,
+        possible_routes,
+        payer_channel,
+        channelmap,
+        payer_transfer,
+        block_number,
+    )
+
+    send_transfer = must_contain_entry(initial_iteration.events, SendMediatedTransfer, {})
+    assert send_transfer
+
+    lock_expiration = send_transfer.transfer.lock.expiration
+
+    new_state = initial_iteration.new_state
+    for block_number in range(block_number, lock_expiration + 1):
+        block_state_change = Block(block_number)
+
+        block_iteration = mediator.handle_block(
+            channelmap,
+            new_state,
+            block_state_change,
+            block_number,
+        )
+        new_state = block_iteration.new_state
+
+    # send the balance proof, transitioning the payee state to payed
+    assert new_state.transfers_pair[0].payee_state == 'payee_pending'
+    receive_secret = ReceiveSecretReveal(
+        UNIT_SECRET,
+        channel1.partner_state.address,
+    )
+    payed_iteration = mediator.state_transition(
+        new_state,
+        receive_secret,
+        channelmap,
+        block_number,
+    )
+    payed_state = payed_iteration.new_state
+    assert payed_state.transfers_pair[0].payee_state == 'payee_balance_proof'
+
+    # move to the block in which the payee lock expires, this must not raise an
+    # assertion
+    expired_block_number = lock_expiration + 1
+    expired_block_state_change = Block(expired_block_number)
+    block_iteration = mediator.handle_block(
+        channelmap,
+        payed_state,
+        expired_block_state_change,
+        expired_block_number,
+    )

--- a/raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py
+++ b/raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py
@@ -18,7 +18,7 @@ from raiden.tests.utils.factories import (
 
 
 def test_payer_enter_danger_zone_with_transfer_payed():
-    """ A mediator may have paid the next hop (payee), and didn't get payed by
+    """ A mediator may have paid the next hop (payee), and didn't get paid by
     the previous hop (payer).
 
     When this happens, an assertion must not be hit, because it means the
@@ -83,28 +83,28 @@ def test_payer_enter_danger_zone_with_transfer_payed():
         )
         new_state = block_iteration.new_state
 
-    # send the balance proof, transitioning the payee state to payed
+    # send the balance proof, transitioning the payee state to paid
     assert new_state.transfers_pair[0].payee_state == 'payee_pending'
     receive_secret = ReceiveSecretReveal(
         UNIT_SECRET,
         channel1.partner_state.address,
     )
-    payed_iteration = mediator.state_transition(
+    paid_iteration = mediator.state_transition(
         new_state,
         receive_secret,
         channelmap,
         block_number,
     )
-    payed_state = payed_iteration.new_state
-    assert payed_state.transfers_pair[0].payee_state == 'payee_balance_proof'
+    paid_state = paid_iteration.new_state
+    assert paid_state.transfers_pair[0].payee_state == 'payee_balance_proof'
 
-    # move to the block in which the payee lock expires, this must not raise an
+    # move to the block in which the payee lock expires. This must not raise an
     # assertion
     expired_block_number = lock_expiration + 1
     expired_block_state_change = Block(expired_block_number)
     block_iteration = mediator.handle_block(
         channelmap,
-        payed_state,
+        paid_state,
         expired_block_state_change,
         expired_block_number,
     )

--- a/raiden/transfer/mediated_transfer/mediator.py
+++ b/raiden/transfer/mediated_transfer/mediator.py
@@ -465,9 +465,7 @@ def set_payee_state_and_check_reveal_order(  # pylint: disable=invalid-name
 
 
 def set_expired_pairs(transfers_pair, block_number):
-    """ Set the state transfers to the expired state and return the failed
-    events.
-    """
+    """ Set the state transfers to the expired state and return the failed events."""
     pending_transfers_pairs = get_pending_transfer_pairs(transfers_pair)
 
     events = list()
@@ -484,7 +482,7 @@ def set_expired_pairs(transfers_pair, block_number):
         if has_payer_transfer_expired:
             # For safety, the correct behavior is:
             #
-            # - If the payee has been payed, then the payer must pay too.
+            # - If the payee has been paid, then the payer must pay too.
             #
             #   And the corollary:
             #
@@ -493,7 +491,7 @@ def set_expired_pairs(transfers_pair, block_number):
             #
             # The problem is that this corollary cannot be asserted. If a user
             # is running Raiden without a monitoring service, then it may go
-            # offline after having payed a transfer to a payee, but without
+            # offline after having paid a transfer to a payee, but without
             # getting a balance proof of the payer, and once it comes back
             # online the transfer may have expired.
             #
@@ -533,8 +531,7 @@ def events_for_refund_transfer(refund_channel, refund_transfer, timeout_blocks, 
         block_number (int): The current block number.
     Returns:
         An empty list if there are not enough blocks to safely create a refund,
-        or a list with a refund event.
-    """
+        or a list with a refund event."""
     # A refund transfer works like a special SendMediatedTransfer, so it must
     # follow the same rules and decrement reveal_timeout from the
     # payee_transfer.

--- a/raiden/transfer/mediated_transfer/mediator.py
+++ b/raiden/transfer/mediated_transfer/mediator.py
@@ -465,37 +465,56 @@ def set_payee_state_and_check_reveal_order(  # pylint: disable=invalid-name
 
 
 def set_expired_pairs(transfers_pair, block_number):
-    """ Set the state of expired transfers, and return the failed events. """
+    """ Set the state transfers to the expired state and return the failed
+    events.
+    """
     pending_transfers_pairs = get_pending_transfer_pairs(transfers_pair)
 
     events = list()
     for pair in pending_transfers_pairs:
-        if block_number > pair.payer_transfer.lock.expiration:
-            assert pair.payee_state == 'payee_expired'
-            assert pair.payee_transfer.lock.expiration < pair.payer_transfer.lock.expiration
+        has_payee_transfer_expired = (
+            block_number > pair.payee_transfer.lock.expiration and
+            pair.payee_state != 'payee_expired'
+        )
+        has_payer_transfer_expired = (
+            block_number > pair.payer_transfer.lock.expiration and
+            pair.payer_state != 'payer_expired'
+        )
 
-            if pair.payer_state != 'payer_expired':
-                pair.payer_state = 'payer_expired'
-                # XXX: emit the event only once
-                withdraw_failed = EventWithdrawFailed(
-                    pair.payer_transfer.identifier,
-                    pair.payer_transfer.lock.hashlock,
-                    'lock expired',
-                )
-                events.append(withdraw_failed)
+        if has_payer_transfer_expired:
+            # For safety, the correct behavior is:
+            #
+            # - If the payee has been payed, then the payer must pay too.
+            #
+            #   And the corollary:
+            #
+            # - If the payer transfer has expired, then the payee transfer must
+            #   have expired too.
+            #
+            # The problem is that this corollary cannot be asserted. If a user
+            # is running Raiden without a monitoring service, then it may go
+            # offline after having payed a transfer to a payee, but without
+            # getting a balance proof of the payer, and once it comes back
+            # online the transfer may have expired.
+            #
+            # assert pair.payee_state == 'payee_expired'
 
-        elif block_number > pair.payee_transfer.lock.expiration:
-            assert pair.payee_state not in STATE_TRANSFER_PAID
-            assert pair.payee_transfer.lock.expiration < pair.payer_transfer.lock.expiration
+            pair.payer_state = 'payer_expired'
+            withdraw_failed = EventWithdrawFailed(
+                pair.payer_transfer.identifier,
+                pair.payer_transfer.lock.hashlock,
+                'lock expired',
+            )
+            events.append(withdraw_failed)
 
-            if pair.payee_state != 'payee_expired':
-                pair.payee_state = 'payee_expired'
-                unlock_failed = EventUnlockFailed(
-                    pair.payee_transfer.identifier,
-                    pair.payee_transfer.lock.hashlock,
-                    'lock expired',
-                )
-                events.append(unlock_failed)
+        elif has_payee_transfer_expired:
+            pair.payee_state = 'payee_expired'
+            unlock_failed = EventUnlockFailed(
+                pair.payee_transfer.identifier,
+                pair.payee_transfer.lock.hashlock,
+                'lock expired',
+            )
+            events.append(unlock_failed)
 
     return events
 
@@ -870,6 +889,8 @@ def handle_block(channelidentifiers_to_channels, state, state_change, block_numb
     Return:
         TransitionResult: The resulting iteration
     """
+    assert state_change.block_number == block_number
+
     close_events = events_for_close(
         channelidentifiers_to_channels,
         state.transfers_pair,

